backup: simple NAS backup plugin for KVM

rohityadavcloud opened the pull request:
This is a simple NAS backup plugin for KVM which may be later expanded for other hypervisors. This backup plugin aims to use shared NAS storage on KVM hosts such as NFS (or CephFS and others in future), which is used to backup fully cloned VMs for backup & restore operations. This may NOT be as efficient and performant as some of the other B&R providers, but maybe useful for some KVM environments who are okay to only have full-instance backups and limited functionality.

Design & Implementation follows the `networker` B&R plugin, which is simply:

- Implement B&R plugin interfaces
- Use cmd-answer pattern to execute backup and restore operations on KVM host when VM is running (or needs to be restored) - instead of a B&R API client, relies on answers from KVM agent which executes the operations
- Backups are full VM domain snapshots, copied to a VM-specific folders on a NAS target (NFS) along with a domain XML
- Backup uses libvirt feature: https://libvirt.org/kbase/live_full_disk_backup.html orchestrated via virsh/bash script (nasbackup.sh) as the libvirt-java lacks the bindings
- Supported instance volume storage for restore operations: NFS & local storage

Refer the doc PR for feature limitations and usage details:
https://github.com/apache/cloudstack-documentation/pull/429

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Screenshots

Each backup offering map to a Backup repository that takes in backup storage details (address/url, credentials/mount options):
<img width="1278" alt="Screenshot 2024-08-02 at 5 32 41 PM" src="https://github.com/user-attachments/assets/1b48ddbd-6e42-40e7-bd3c-16b42cc40da3">

Rest is same as B&R feature works; we assign such offering to a VM and backup once completed shows up in the list:
<img width="979" alt="Screenshot 2024-08-02 at 5 58 49 PM" src="https://github.com/user-attachments/assets/dcf98b3c-9386-4555-8654-ef6d7ce2c47c">
<img width="1598" alt="Screenshot 2024-08-02 at 5 59 07 PM" src="https://github.com/user-attachments/assets/8cbe0128-8350-47d9-b5c1-3257299155e4">

rohityadavcloud commented:
Update: take backup & delete backup implemented for NFS; TBD/TODO more operations.

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10548

rohityadavcloud commented:
Tested NFS, CephFS & samba, works when configuration is correctly configured; for example;

```
mysql> select * from backup_repository\G;
*************************** 1. row ***************************
            id: 1
          uuid: d3b911d4-50a6-11ef-bc8b-525400f93984
          name: NFS-backup
       zone_id: 1
      provider: nas
          type: nfs
       address: 172.20.0.1:/export/testing/qa1-cs420-kvm-1cdf7971-kvm/backup
    mount_opts:
    used_bytes: NULL
capacity_bytes: NULL
       created: 2024-08-02 08:11:31
       removed: NULL
*************************** 2. row ***************************
            id: 2
          uuid: 09c2efd1-50c4-11ef-bc8b-525400f93984
          name: Ceph-backup
       zone_id: 1
      provider: nas
          type: ceph
       address: 10.10.1.10,10.10.1.11,10.10.1.12:/backup
    mount_opts: name=rohit,secret=XXXX==
    used_bytes: NULL
capacity_bytes: NULL
       created: 2024-08-02 11:40:37
       removed: NULL
*************************** 3. row ***************************
            id: 3
          uuid: 876b1da4-50c6-11ef-bc8b-525400f93984
          name: Samba-backup
       zone_id: 1
      provider: nas
          type: cifs
       address: //192.168.1.10/backup
    mount_opts: vers=3.0,username=rohit,password=XXXX
    used_bytes: NULL
capacity_bytes: NULL
       created: 2024-08-02 11:58:27
       removed: NULL
3 rows in set (0.00 sec)
```

rohityadavcloud commented:
Shared storage mount/umount tested:
- EL8 & EL9: NFS & CephFS works (ceph would require rbd utils, otherwise uses Linux kernel's cephfs driver)
- Ubuntu: NFS, CephFS & CIFS/SMB3 works (the latter two need cifs-utils & ceph utils)

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10579

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10580

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10583

DaanHoogland commented:
nice feature idea @rohityadavcloud :+1:

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10621

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10624

rohityadavcloud commented on tools/marvin/setup.py:
```diff
@@ -27,7 +27,7 @@
 
 
-VERSION = "4.20.0.0-SNAPSHOT"
+VERSION = "4.20.0.0"
```
```suggestion
VERSION = "4.20.0.0-SNAPSHOT"
```

DaanHoogland replied:
I think this change should not be here, but also this is easy to miss. should it be removed?

rohityadavcloud commented on ui/src/views/AutogenView.vue:
```diff
@@ -1174,6 +1174,7 @@ export default {
 
       this.showAction = true
       const listIconForFillValues = ['copy-outlined', 'CopyOutlined', 'edit-outlined', 'EditOutlined', 'share-alt-outlined', 'ShareAltOutlined']
+      console.log(this.currentAction.paramFields)
```
```suggestion
```

rohityadavcloud commented:
Update cc @DaanHoogland @sureshanaparti @shwstppr @rajujith @NuxRo the experimental plugin is ready for review & testing, pending a bit of cleanup, maybe tests & doc work. Tested following cases:

- Backup of VM with root disk & VM with root & data disks
- Restore of backup of VM with root disk & VM with root & data disks
- Deleting backups of VM with root disk & VM with root & data disks
- Restore a volume of a backup of VM

Case WIP: restoring VM using backup for a VM which is expunged

@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

andrijapanicsb commented:
@Pearl1594 do we have more (internal?) details on the design of this one? Looks really interesting.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10847

rohityadavcloud commented:
There isn't a design doc or spec, I'll submit a doc PR which can be used for notes. The plugin simply uses existing B&R framework and follows on the footsteps of `networker` plugin, so its well isolated as yet another plugin - all supported B&R functionality work as-is, doc PR will make it further clear. 

@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

shwstppr commented on server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:
```diff
@@ -749,20 +759,26 @@ public boolean restoreBackupVolumeAndAttachToVM(final String backedUpVolumeUuid,
+        BackupProvider backupProvider = getBackupProvider(offering.getProvider());
+        VolumeVO backedUpVolume = volumeDao.findByUuid(backedUpVolumeUuid);
+        Pair<HostVO, StoragePoolVO> restoreInfo;
+        if (!Objects.equals(offering.getProvider(), "nas")) {
```
```suggestion
        if (!"nas".equals(offering.getProvider())) {
```

shwstppr commented on ui/src/components/view/ListResourceTable.vue:
```diff
@@ -50,6 +50,10 @@
             {{ $toLocaleDate(text) }}
           </template>
 
+          <template v-else-if="column.key === 'size' || column.key === 'virtualsize'">
+            {{ parseFloat(text / (1024.0 * 1024.0 * 1024.0)).toFixed(2) }} GB
+          </template>
+
```
would it be better to use - `bytesToHumanReadableSize` method?

rohityadavcloud replied:
@Pearl1594 to check & advise.

shwstppr commented on ui/src/components/view/ListView.vue:
```diff
@@ -174,7 +174,7 @@
         <span>{{ text <= 0 || !text ? 'N/A' : text }}</span>
       </template>
-      <template v-else-if="column.key === 'size'">
+      <template v-else-if="column.key === 'size' || column.key === 'virtualsize'">
```
minor (presentational)
```suggestion
      <template v-else-if="['size', 'virtualsize'].includes(column.key)">
```

rohityadavcloud commented on plugins/backup/nas/src/main/java/org/apache/cloudstack/backup/NASBackupProvider.java:
```diff
@@ -0,0 +1,434 @@
+        command.setBackupRepoAddress(backupRepository.getAddress());
+        command.setMountOptions(backupRepository.getMountOptions());
+
+        if (VirtualMachine.State.Stopped.equals(vm.getState())) {
+            List<VolumeVO> vmVolumes = volumeDao.findByInstance(vm.getId());
+            List<String> volumePaths = getVolumePaths(vmVolumes);
+            command.setVolumePaths(volumePaths);
+        }
```
ISO backups are not considered.

blueorangutan commented:
<b>[SF] Trillian test result (tid-11228)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 49421 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9451-t11228-kvm-ol8.zip
Smoke tests completed. 139 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

rajujith commented:
@blueorangutan  package

rajujith commented:
@blueorangutan package

blueorangutan commented:
@rajujith a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

shwstppr commented on server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:
```diff
@@ -23,6 +23,7 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.TimeZone;
```
```suggestion
import java.util.TimeZone;
```

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

shwstppr commented on core/src/main/java/org/apache/cloudstack/backup/DeleteBackupCommand.java:
```diff
@@ -0,0 +1,76 @@
+    public boolean executeInSequence() {
+        return true;
+    }
+}
```
GH action Lint error
```suggestion
}

```

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10876

rajujith commented:
@blueorangutan test

blueorangutan commented:
@rajujith a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

DaanHoogland reviewed (commented):
code looks good, but I have some questions about the code organisation;
1. the code is kvm specific? should more code be moved to the KVM plugin?
2. the backup repository is specific to the nas provider but most of the classes are in generic packages like api, client, engine, server. Should more be moved to the plugins?

rohityadavcloud commented on core/src/main/java/org/apache/cloudstack/backup/TakeBackupCommand.java:
```diff
@@ -0,0 +1,94 @@
+
+import java.util.List;
+
+public class TakeBackupCommand extends Command {
```
There's are take snapshot related method/classes, so we followed that naming convention. CreateBackup name also makes sense.

rohityadavcloud commented:
Thanks @shwstppr @DaanHoogland for the review, let me try & address as much as I could.

@DaanHoogland to answer your comments;

1. yes, this is mostly for KVM; however, we think this may be in future expanded to other hypervisors; as most hypervisors support NFS (at least) among the supported NAS types; the method to take/restore backup could use full VM snapshots of non-KVM hypervisors.
2. yes, backup repository concept for now is specific to only this plugin - however, if the scope expands to a general purpose B&R plugin, then this becomes non-specific to the plugin. Also I think the networkers & veeam plugins have some concept of backup repository or backup storage which can reuse this concept. That said, I'm happy to move the dao/vo logic to the plugin if you really insist (and refactor in future).

rohityadavcloud commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRestoreBackupCommandWrapper.java:
```diff
@@ -0,0 +1,181 @@
+        String bkpPath = String.format(FILE_PATH_PLACEHOLDER, mountDirectory, backupPath);
+        int lastIndex = volumePath.lastIndexOf("/");
+        String volUuid = Objects.isNull(volumeUuid) ? volumePath.substring(lastIndex + 1) : volumeUuid;
+        String backupFileName = String.format("%s.%s.%s.qcow2", deviceId, diskType.toLowerCase(Locale.ROOT), volUuid);
```
That's a limitation of the B&R plugin, it will be always qcow2 - see docs at https://github.com/apache/cloudstack-documentation/pull/429 - this can change in future.

DaanHoogland commented:
> Thanks @shwstppr @DaanHoogland for the review, let me try & address as much as I could.
> 
> @DaanHoogland to answer your comments;
> 
>     1. yes, this is mostly for KVM; however, we think this may be in future expanded to other hypervisors; as most hypervisors support NFS (at least) among the supported NAS types; the method to take/restore backup could use full VM snapshots of non-KVM hypervisors.

ok, great.

>     2. yes, backup repository concept for now is specific to only this plugin - however, if the scope expands to a general purpose B&R plugin, then this becomes non-specific to the plugin. Also I think the networkers & veeam plugins have some concept of backup repository or backup storage which can reuse this concept. That said, I'm happy to move the dao/vo logic to the plugin if you really insist (and refactor in future).

If it is that's fine. It was just not that obvious to me.

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10888

blueorangutan commented:
<b>[SF] Trillian test result (tid-11248)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 49729 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9451-t11248-kvm-ol8.zip
Smoke tests completed. 139 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

Pearl1594 commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRestoreBackupCommandWrapper.java:
```diff
@@ -0,0 +1,181 @@
+                try {
+                    replaceVolumeWithBackup(volumePath, bkpPathAndVolUuid.first());
+                } catch (IOException e) {
+                    throw new CloudRuntimeException(String.format("Unable to revert backup for volume [%s] due to [%s].", bkpPathAndVolUuid.second(), e.getMessage()), e);
```
Nothing as of now. One can re-attempt again.

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10916

Pearl1594 commented:
@blueorangutan test

blueorangutan commented:
@Pearl1594 a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-11304)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 49496 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9451-t11304-kvm-ol8.zip
Smoke tests completed. 139 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10934

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10935

Pearl1594 commented:
@blueorangutan test

blueorangutan commented:
@Pearl1594 a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-11323)<b/>

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rohityadavcloud commented on plugins/backup/nas/src/main/java/org/apache/cloudstack/backup/NASBackupProvider.java:
```diff
@@ -0,0 +1,434 @@
+    protected Host getLastVMHypervisorHost(VirtualMachine vm) {
+        Long hostId = vm.getLastHostId();
+        if (hostId == null) {
+            LOG.debug("Cannot find last host for vm. This should never happen, please check your database.");
```
@Pearl1594 to check & advise.

rohityadavcloud commented on plugins/backup/nas/src/main/java/org/apache/cloudstack/backup/NASBackupProvider.java:
```diff
@@ -0,0 +1,434 @@
+            throw new CloudRuntimeException("No valid backup repository found for the VM, please check the attached backup offering");
+        }
+
+        // TODO: this can be any host in the cluster or last host
```
@Pearl1594 to check & advise.

rohityadavcloud commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRestoreBackupCommandWrapper.java:
```diff
@@ -0,0 +1,181 @@
+        try {
+            mountDirectory = Files.createTempDirectory(mountDirectory).toString();
+            String mount = String.format(MOUNT_COMMAND, backupRepoType, backupRepoAddress, mountDirectory);
+            Script.runSimpleBashScript(mount);
```
@Pearl1594 can you check, if that should be used?

Pearl1594 replied:
Will bear in mind this improvement as part of an improvement PR that will be raised.

rohityadavcloud commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRestoreBackupCommandWrapper.java:
```diff
@@ -0,0 +1,181 @@
+    private void unmountBackupDirectory(String backupDirectory) {
+        try {
+            String umountCmd = String.format(UMOUNT_COMMAND, backupDirectory);
+            Script.runSimpleBashScript(umountCmd);
```
@Pearl1594 to check & advise.

rohityadavcloud commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRestoreBackupCommandWrapper.java:
```diff
@@ -0,0 +1,181 @@
+            for (int i = 0; i < volumePaths.size(); i++) {
+                String volumePath = volumePaths.get(i);
+                Pair<String, String> bkpPathAndVolUuid = getBackupPath(mountDirectory, volumePath, backupPath, diskType, i, null);
+                diskType = "datadisk";
```
@Pearl1594 to check & advise.

Pearl1594 replied:
will be addressed as an improvement.

shwstppr commented on api/src/main/java/org/apache/cloudstack/api/command/user/backup/ListBackupScheduleCmd.java:
```diff
@@ -74,9 +79,14 @@ public Long getVmId() {
+            List<BackupSchedule> schedules = backupManager.listBackupSchedule(getVmId());
+            ListResponse<BackupScheduleResponse> response = new ListResponse<>();
+            List<BackupScheduleResponse> scheduleResponses = new ArrayList<>();
+            if (Objects.nonNull(schedules) && !schedules.isEmpty()) {
```
CollectionUtils?

shwstppr commented on api/src/main/java/org/apache/cloudstack/backup/Backup.java:
```diff
@@ -141,5 +142,6 @@ public String toString() {
     Backup.Status getStatus();
     Long getSize();
     Long getProtectedSize();
+    List<VolumeInfo> getBackedVolumes();
```
minor - getBackedUpVolumes?

shwstppr commented on engine/schema/src/main/java/org/apache/cloudstack/backup/BackupVO.java:
```diff
@@ -82,6 +87,9 @@ public class BackupVO implements Backup {
     @Column(name = "zone_id")
     private long zoneId;
 
+    @Column(name = "backed_volumes", length = 65535)
+    protected String backedVolumes;
```
What does this store? UUID or path? Would it have been better to use a details or backup_volume_map table?

Pearl1594 replied:
I've followed the convention of the backup framework. A similar column is present in the vm_instance table.

shwstppr commented on engine/schema/src/main/java/org/apache/cloudstack/backup/dao/BackupDaoImpl.java:
```diff
@@ -161,7 +162,14 @@ public BackupResponse newBackupResponse(Backup backup) {
         response.setProtectedSize(backup.getProtectedSize());
         response.setStatus(backup.getStatus());
-        response.setVolumes(new Gson().toJson(vm.getBackupVolumeList().toArray(), Backup.VolumeInfo[].class));
+        // ACS 4.20: For backups taken prior this release the backup.backed_volumes column would be empty hence use vm_instance.backup_volumes
```
Understood the use `backup.backed_volumes` wrt my preivous comment.

shwstppr reviewed (approved):
code lgtm

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10951

Pearl1594 commented on ui/src/views/compute/backup/BackupSchedule.vue:
```diff
@@ -147,18 +147,15 @@ export default {
     this.dataSchedules = []
     if (this.dataSource && Object.keys(this.dataSource).length > 0) {
-      this.dataSchedules.push(this.dataSource)
+      this.dataSchedules = this.dataSource
```
Yes, that's right. Earlier, i.e before this PR, backup schedules was always 1 per instance due to a limitation (https://github.com/apache/cloudstack/issues/9344) . But now that it is a list, this would cause issues.

blueorangutan commented:
<b>[SF] Trillian test result (tid-11318)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 54306 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9451-t11318-kvm-ol8.zip
Smoke tests completed. 138 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_secure_vm_migration | `Error` | 306.60 | test_vm_life_cycle.py
test_02_unsecure_vm_migration | `Error` | 476.94 | test_vm_life_cycle.py
test_03_secured_to_nonsecured_vm_migration | `Error` | 379.11 | test_vm_life_cycle.py

Pearl1594 commented:
@blueorangutan package

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
@Pearl1594 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10971

Pearl1594 commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10972

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10978

rajujith commented:
@blueorangutan  package

rajujith commented:
@blueorangutan package

blueorangutan commented:
@rajujith a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 10986

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

sureshanaparti commented on engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql:
```diff
@@ -234,6 +254,15 @@ CALL `cloud`.`IDEMPOTENT_MODIFY_COLUMN_CHAR_SET`('vpc_offerings', 'display_text'
+
+-- Add volume details to the backups table to keep track of the volumes being backed up
+CALL `cloud`.`IDEMPOTENT_ADD_COLUMN`('cloud.backups', 'backed_volumes', 'text DEFAULT NULL COMMENT "details of backed-up volumes" ');
+
```
```suggestion
CALL `cloud`.`IDEMPOTENT_MODIFY_COLUMN_CHAR_SET`('backups', 'backed_volumes', 'TEXT', 'DEFAULT NULL COMMENT \'details of backed-up volumes\'')
```

add this, to support unicode chars for the backed up volume details

sureshanaparti commented on engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql:
```diff
@@ -141,6 +141,26 @@ CREATE TABLE `cloud`.`webhook_delivery` (
+CREATE TABLE `cloud`.`backup_repository` (
+  `id` bigint unsigned NOT NULL AUTO_INCREMENT COMMENT 'id of the backup repository',
+  `uuid` varchar(255) NOT NULL COMMENT 'uuid of the backup repository',
+  `name` varchar(255) NOT NULL COMMENT 'name of the backup repository',
```
```suggestion
  `name` varchar(255) CHARACTER SET utf8mb4 NOT NULL COMMENT 'name of the backup repository',
```

to support unicode chars in the name

sureshanaparti commented on plugins/backup/nas/src/main/java/org/apache/cloudstack/backup/NASBackupProvider.java:
```diff
@@ -0,0 +1,442 @@
+
+    @Override
+    public String getName() {
+        return "nas";
```
use string constant, also use the same in NasBackupOffering.java , etc

rohityadavcloud replied:
All other providers export string name, constants aren't required because of this. We can check maybe introduce them like enums if necessary. (same pattern used across several other types of plugins)

sureshanaparti commented on server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:
```diff
@@ -945,6 +985,9 @@ public List<Class<?>> getCommands() {
         cmdList.add(RestoreVolumeFromBackupAndAttachToVMCmd.class);
+        cmdList.add(AddBackupRepositoryCmd.class);
+        cmdList.add(DeleteBackupRepositoryCmd.class);
+        cmdList.add(ListBackupRepositoriesCmd.class);
```
Update cmd required? to update name or mount options?

rohityadavcloud replied:
@Pearl1594 maybe we can export/enable the APIs only when nas plugin is enabled?

rohityadavcloud replied:
Actually I double checked, plugins are configureable even on a per zone level - so we'll need to export all the APIs when the feature is enabled.

sureshanaparti reviewed (approved):
minor comments, overall code looks good.

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rajujith reviewed (approved):
LGTM

- Works with KVM EL8 ,EL9, Ubuntu22,Ubuntu24 
- Backup and restore works fine. All Full backups. 
- Supported backup repository NFS, tested using QNAP. 
- Parallel backups and restore
- Parallel backups and restore per host.
- Multiple Linux distros and Windows.
- Multiple Backup repositories.
- Backup schedules.
- Admin and users can trigger backups.
- Primary storages of type NFS and local storage. 
- Upgrade from 4.19.


Limitations observed and possible improvements( some are limitations of the B&R Framework itself)

- Restoring expunged instances.
- Restoring unmanaged instances.
- More backup repository type support.
- CKS support
- Agent logging could be improved.

rohityadavcloud commented:
@blueorangutan package

Pearl1594 commented:
Few comments will be addressed as improvements in a subsequent PR.

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

rohityadavcloud commented:
Merging this based on reviews, manual QA and GitHub Actions tests. Will continue further test matrix on health check PR.

blueorangutan commented:
<b>[SF] Trillian test result (tid-11369)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 53228 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9451-t11369-kvm-ol8.zip
Smoke tests completed. 138 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_secure_vm_migration | `Error` | 135.16 | test_vm_life_cycle.py
test_01_secure_vm_migration | `Error` | 135.16 | test_vm_life_cycle.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-11377)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 54143 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9451-t11377-kvm-ol8.zip
Smoke tests completed. 139 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_mount_shared_fs | `Failure` | 84.98 | test_sharedfs_lifecycle.py
test_resize_shared_fs | `Error` | 0.44 | test_sharedfs_lifecycle.py
